### PR TITLE
[ADD] Documentation for data validation and preprocessing

### PR DESCRIPTION
**Pull request by ravinkohli**
Addresses #319 .
Also, fixes formatting issues which were caused by improper indentation.

### CODE REVIEW

File: autoPyTorch/api/base_task.py [hunk 1]
```diff
@@ -102,13 +102,16 @@ def send_warnings_to_log(
 class BaseTask:
     """
     Base class for the tasks that serve as API to the pipelines.
+
     Args:
-        seed (int), (default=1): seed to be used for reproducibility.
-        n_jobs (int), (default=1): number of consecutive processes to spawn.
+        seed (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:107]: This looks better? (I could not find an example of default in google doc-string)
```suggestion
        seed (int: default=1):
```

File: autoPyTorch/api/base_task.py [hunk 2]
```diff
@@ -1317,10 +1330,16 @@ def get_incumbent_results(
     ) -> Tuple[Configuration, Dict[str, Union[int, str, float]]]:
         """
         Get Incumbent config and the corresponding results
+
         Args:
-            include_traditional: Whether to include results from tradtional pipelines
+            include_traditional:
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:1335]:
```suggestion
            include_traditional (bool):
```

File: autoPyTorch/api/base_task.py [hunk 3]
```diff
@@ -1317,10 +1330,16 @@ def get_incumbent_results(
...
+                Whether to include results from tradtional pipelines
 
         Returns:
+            Configuration:
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:1339]:
```suggestion
            Configuration (CS.ConfigurationSpace):
```

File: autoPyTorch/api/base_task.py [hunk 4]
```diff
@@ -1354,6 +1373,13 @@ def get_models_with_weights(self) -> List:
         return models_with_weights
 
     def show_models(self) -> str:
+        """
+        Returns a Markdown containing details about the final ensemble/configuration.
+
+        Returns:
+            str:
+                Markdown table.
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:1381]:
```suggestion
                Markdown table of models.
```

**Review by nabenabe0928 — COMMENTED** [review]
Thanks for the PR.
I put comments except `docs/manual.rst` .

**Comment by nabenabe0928** [comment]
docs/manual.rst looks good to me, so I will wait for your commits.

File: autoPyTorch/api/base_task.py [hunk 5]
```diff
@@ -102,13 +102,16 @@ def send_warnings_to_log(
 class BaseTask:
     """
     Base class for the tasks that serve as API to the pipelines.
+
     Args:
-        seed (int), (default=1): seed to be used for reproducibility.
-        n_jobs (int), (default=1): number of consecutive processes to spawn.
+        seed (int: default=1):
+            seed to be used for reproducibility.
+        n_jobs (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:109]:
```suggestion
        n_jobs (int: default=1):
```

File: autoPyTorch/api/base_task.py [hunk 6]
```diff
@@ -102,13 +102,16 @@ def send_warnings_to_log(
...
+            number of consecutive processes to spawn.
         n_threads (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:111]:
```suggestion
        n_threads (int: default=1):
```

File: autoPyTorch/api/base_task.py [hunk 7]
```diff
@@ -102,13 +102,16 @@ def send_warnings_to_log(
...
             number of threads to use for each process.
-        logging_config (Optional[Dict]): specifies configuration
-            for logging, if None, it is loaded from the logging.yaml
+        logging_config (Optional[Dict]):
+            Specifies configuration for logging, if None, it is loaded from the logging.yaml
         ensemble_size (int), (default=50): Number of models added to the ensemble built by
```
Comment (nabenabe0928) [inline autoPyTorch/api/base_task.py:115]:
```suggestion
        ensemble_size (int: default=50): Number of models added to the ensemble built by
```

File: autoPyTorch/api/tabular_classification.py [hunk 8]
```diff
@@ -25,37 +25,42 @@
 class TabularClassificationTask(BaseTask):
     """
     Tabular Classification API to the pipelines.
+
     Args:
-        seed (int), (default=1): seed to be used for reproducibility.
-        n_jobs (int), (default=1): number of consecutive processes to spawn.
+        seed (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:30]:
```suggestion
        seed (int: default=1):
```

File: autoPyTorch/api/tabular_classification.py [hunk 9]
```diff
@@ -25,37 +25,42 @@
...
+            seed to be used for reproducibility.
+        n_jobs (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:32]:
```suggestion
        n_jobs (int: default=1):
```

File: autoPyTorch/api/tabular_classification.py [hunk 10]
```diff
@@ -25,37 +25,42 @@
...
+            number of consecutive processes to spawn.
         n_threads (int), (default=1):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:34]:
```suggestion
        n_threads (int: default=1):
```

File: autoPyTorch/api/tabular_classification.py [hunk 11]
```diff
@@ -25,37 +25,42 @@
...
             number of threads to use for each process.
         logging_config (Optional[Dict]):
-            specifies configuration for logging, if None, it is loaded from the logging.yaml
+            Specifies configuration for logging, if None, it is loaded from the logging.yaml
         ensemble_size (int), (default=50):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:38]:
```suggestion
        ensemble_size (int: default=50):
```

File: autoPyTorch/api/tabular_classification.py [hunk 12]
```diff
@@ -25,37 +25,42 @@
...
             Number of models added to the ensemble built by
             Ensemble selection from libraries of models.
             Models are drawn with replacement.
         ensemble_nbest (int), (default=50):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:42]:
```suggestion
        ensemble_nbest (int: default=50):
```

File: autoPyTorch/api/tabular_classification.py [hunk 13]
```diff
@@ -25,37 +25,42 @@
...
-            only consider the ensemble_nbest
+            Only consider the ensemble_nbest
             models to build the ensemble
         max_models_on_disc (int), (default=50):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:45]:
```suggestion
        max_models_on_disc (int: default=50):
```

File: autoPyTorch/api/tabular_classification.py [hunk 14]
```diff
@@ -174,8 +189,8 @@ def search(
                 in seconds for the search of appropriate models.
                 By increasing this value, autopytorch has a higher
                 chance of finding better models.
-            func_eval_time_limit_secs (int), (default=None):
-                Time limit for a single call to the machine learning model.
+            func_eval_time_limit_secs (int), (default=None): Time limit
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:192]:
```suggestion
            func_eval_time_limit_secs (int: default=None): Time limit
```

File: autoPyTorch/api/tabular_classification.py [hunk 15]
```diff
@@ -188,35 +203,40 @@ def search(
             enable_traditional_pipeline (bool), (default=True):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:203]:
```suggestion
            enable_traditional_pipeline (bool: default=True):
```

File: autoPyTorch/api/tabular_classification.py [hunk 16]
```diff
@@ -188,35 +203,40 @@ def search(
...
                 We fit traditional machine learning algorithms
                 (LightGBM, CatBoost, RandomForest, ExtraTrees, KNN, SVM)
-                before building PyTorch Neural Networks. You can disable this
+                prior building PyTorch Neural Networks. You can disable this
                 feature by turning this flag to False. All machine learning
                 algorithms that are fitted during search() are considered for
                 ensemble building.
-            memory_limit (Optional[int]), (default=4096):
-                Memory limit in MB for the machine learning algorithm. autopytorch
+            memory_limit (Optional[int]), (default=4096): Memory
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:210]:
```suggestion
            memory_limit (Optional[int]: default=4096): Memory
```

File: autoPyTorch/api/tabular_classification.py [hunk 17]
```diff
@@ -188,35 +203,40 @@ def search(
...
                 instances, num_params, runhistory, seed and ta. This is
                 an advanced feature. Use only if you are familiar with
                 [SMAC](https://automl.github.io/SMAC3/master/index.html).
-            all_supported_metrics (bool), (default=True):
-                if True, all metrics supporting current task will be calculated
+            tae_func (Optional[Callable]):
+                TargetAlgorithm to be optimised. If None, `eval_function`
+                available in autoPyTorch/evaluation/train_evaluator is used.
+                Must be child class of AbstractEvaluator.
+            all_supported_metrics (bool), (default=True): if True, all
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:232]:
```suggestion
            all_supported_metrics (bool: default=True): if True, all
```

File: autoPyTorch/api/tabular_classification.py [hunk 18]
```diff
@@ -188,35 +203,40 @@ def search(
...
+                metrics supporting current task will be calculated
                 for each pipeline and results will be available via cv_results
             precision (int), (default=32): Numeric precision used when loading
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:235]:
```suggestion
            precision (int: default=32): Numeric precision used when loading
```

File: autoPyTorch/api/tabular_classification.py [hunk 19]
```diff
@@ -188,35 +203,40 @@ def search(
...
                 ensemble data. Can be either '16', '32' or '64'.
             disable_file_output (Union[bool, List]):
-            load_models (bool), (default=True):
-                Whether to load the models after fitting AutoPyTorch.
+            load_models (bool), (default=True): Whether to load the
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:238]:
```suggestion
            load_models (bool: default=True): Whether to load the
```

File: autoPyTorch/api/tabular_classification.py [hunk 20]
```diff
@@ -188,35 +203,40 @@ def search(
...
+                models after fitting AutoPyTorch.
             portfolio_selection (str), (default=None):
```
Comment (nabenabe0928) [inline autoPyTorch/api/tabular_classification.py:240]:
```suggestion
            portfolio_selection (str: default=None):
```

**Review by nabenabe0928 — COMMENTED** [review]
I think we can delete `default`?

You can check which file has `(xxx), (default=xxx)` by `find -name "*.py" -exec grep ", (default" {} \; -print | grep "./"`.

I also found out that some of them have the default value of `None` while the typing says `xxx`, but not `Optional[xxx]`.

File: autoPyTorch/api/tabular_classification.py (hunk 14 above)
Comment (ravinkohli) [reply autoPyTorch/api/tabular_classification.py:192]:
```suggestion
            func_eval_time_limit_secs (Optional[int]: default=None): Time limit
```

File: autoPyTorch/api/tabular_classification.py (hunk 16 above)
Comment (ravinkohli) [reply autoPyTorch/api/tabular_classification.py:210]:
```suggestion
            memory_limit (int: default=4096): Memory
```

File: autoPyTorch/api/tabular_classification.py (hunk 20 above)
Comment (ravinkohli) [reply autoPyTorch/api/tabular_classification.py:240]:
```suggestion
            portfolio_selection (Optional[str]):
```

File: autoPyTorch/api/tabular_regression.py [hunk 21]
```diff
@@ -25,30 +25,42 @@
 class TabularRegressionTask(BaseTask):
     """
     Tabular Regression API to the pipelines.
+
     Args:
-        seed (int), (default=1): seed to be used for reproducibility.
-        n_jobs (int), (default=1): number of consecutive processes to spawn.
+        seed (int), (default=1):
```
Comment (ravinkohli) [inline autoPyTorch/api/tabular_regression.py:30]:
```suggestion
        seed (int: default=1):
```

File: autoPyTorch/api/tabular_regression.py [hunk 22]
```diff
@@ -25,30 +25,42 @@
...
+            seed to be used for reproducibility.
+        n_jobs (int), (default=1):
```
Comment (ravinkohli) [inline autoPyTorch/api/tabular_regression.py:32]:
```suggestion
        n_jobs (int: default=1):
```

File: autoPyTorch/api/tabular_regression.py [hunk 23]
```diff
@@ -25,30 +25,42 @@
...
+            number of consecutive processes to spawn.
         n_threads (int), (default=1):
             number of threads to use for each process.
-        logging_config (Optional[Dict]): specifies configuration
-            for logging, if None, it is loaded from the logging.yaml
-        ensemble_size (int), (default=50): Number of models added to the ensemble built by
+        logging_config (Optional[Dict]):
+            Specifies configuration for logging, if None, it is loaded from the logging.yaml
+        ensemble_size (int), (default=50):
```
Comment (ravinkohli) [inline autoPyTorch/api/tabular_regression.py:38]:
```suggestion
        ensemble_size (int: default=50):
```

File: autoPyTorch/api/tabular_regression.py [hunk 24]
```diff
@@ -25,30 +25,42 @@
...
+            Number of models added to the ensemble built by
             Ensemble selection from libraries of models.
             Models are drawn with replacement.
-        ensemble_nbest (int), (default=50): only consider the ensemble_nbest
+        ensemble_nbest (int), (default=50):
```
Comment (ravinkohli) [inline autoPyTorch/api/tabular_regression.py:42]:
```suggestion
        ensemble_nbest (int: default=50):
```

File: autoPyTorch/api/tabular_regression.py [hunk 25]
```diff
@@ -25,30 +25,42 @@
...
+            Only consider the ensemble_nbest
             models to build the ensemble
-        max_models_on_disc (int), (default=50): maximum number of models saved to disc.
-            Also, controls the size of the ensemble as any additional models will be deleted.
+        max_models_on_disc (int), (default=50):
```
Comment (ravinkohli) [inline autoPyTorch/api/tabular_regression.py:45]:
```suggestion
        max_models_on_disc (int: default=50):
```

File: autoPyTorch/api/base_task.py [hunk 26]
```diff
@@ -102,14 +102,17 @@ def send_warnings_to_log(
...
             number of threads to use for each process.
-        logging_config (Optional[Dict]): specifies configuration
-            for logging, if None, it is loaded from the logging.yaml
-        ensemble_size (int), (default=50): Number of models added to the ensemble built by
+        logging_config (Optional[Dict]):
+            Specifies configuration for logging, if None, it is loaded from the logging.yaml
+        ensemble_size (int: default=50): Number of models added to the ensemble built by
             Ensemble selection from libraries of models.
             Models are drawn with replacement.
         ensemble_nbest (int), (default=50): only consider the ensemble_nbest
```
Comment (ravinkohli) [inline autoPyTorch/api/base_task.py:118]:
```suggestion
        ensemble_nbest (int: default=50): only consider the ensemble_nbest
```

File: autoPyTorch/pipeline/components/training/metrics/base.py [hunk 27]
```diff
@@ -96,20 +97,22 @@ def __call__(
...
-        sample_weight : array-like, optional (default=None)
-            Sample weights.
-        Returns
-        -------
-        score : float
-            Score function applied to prediction of estimator on X.
+
+        """
+        Evaluate predicted probabilities for X relative to y_true.
+
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/metrics/base.py:103]:
```suggestion

```

**Comment by ravinkohli** [comment]
> I think we can delete `default`?
> 
> You can check which file has `(xxx), (default=xxx)` by `find -name "*.py" -exec grep ", (default" {} \; -print | grep "./"`.
> 
> I also found out that some of them have the default value of `None` while the typing says `xxx`, but not `Optional[xxx]`.

I have unified the way we document types throughout the project. I have also fixed type information to `Optional[type]` where previously we had default=None. I think if there is any further change required, lets raise an issue for this and fix it in a different PR.

**Pull request merged**